File: examples/js-vuln-demo-no-answers/src/p0_safe_functions.js

const crypto = require('crypto');
const { execFile } = require('child_process');
const path = require('path');


function safeQuery(db, userId) {
    return db.query('SELECT * FROM users WHERE id = ?', [userId]);
}


function generateToken() {
    return crypto.randomBytes(32).toString('hex');
}


function safeExec() {
    execFile('ls', ['-l'], (err, stdout) => {});
}


function logEvent(user) {
    console.log('User login: %s', user);
}


function isSafePath(base, input) {
    const resolved = path.resolve(base, input);
    return resolved.startsWith(path.resolve(base));
}


function safeStringify(data) {
    return JSON.stringify(data);
}

File: examples/java-vuln-demo-no-answers/src/com/secguardian/demo/p0_safe_functions.java
package com.secguardian.demo;


import java.sql.*;
import java.security.*;
import java.nio.file.*;
import java.util.logging.*;

public class p0_safe_functions {
    

    public void safeQuery(Connection conn, int userId) throws SQLException {
        String sql = "SELECT * FROM users WHERE id = ?";
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, userId);
            ResultSet rs = ps.executeQuery();
        }
    }


    public String generateToken() {
        SecureRandom sr = new SecureRandom();
        byte[] token = new byte[32];
        sr.nextBytes(token);
        return bytesToHex(token);
    }


    public void safeMove(Path src, Path dst) throws IOException {
        Files.move(src, dst, StandardCopyOption.ATOMIC_MOVE);
    }


    private static final Logger logger = Logger.getLogger("App");
    public void logEvent(String user) {
        logger.info("User login: {0}", user);  // %s 占位符，非拼接
    }


    public boolean isSafePath(String base, String input) {
        Path resolved = Paths.get(base).resolve(input).normalize();
        return resolved.startsWith(base);
    }


    public String encodeBase64(byte[] data) {
        return Base64.getEncoder().encodeToString(data);
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }
}
